hnix-store-core/src/System/Nix/Internal/Nar/Effects.hs

```diff
@@ -0,0 +1,75 @@
+-- |
+
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE RankNTypes     #-}
+
+module System.Nix.Internal.Nar.Effects
+  ( NarEffects(..)
+  , narEffectsIO
+  ) where
+
+import qualified Control.Monad.IO.Class as IO
+import qualified Data.ByteString        as BS
+import qualified Data.ByteString.Lazy   as BSL
+import           Data.Int               (Int64)
+import qualified System.Directory       as Directory
+import qualified System.IO              as IO
+import           System.Posix.Files     (createSymbolicLink, fileSize,
+                                         getFileStatus, isDirectory,
+                                         readSymbolicLink)
+
+data NarEffects (m :: * -> *) = NarEffects {
+    narReadFile   :: FilePath -> m BSL.ByteString
+  , narWriteFile  :: FilePath -> BSL.ByteString -> m ()
+  , narStreamFile :: FilePath -> m (Maybe BS.ByteString) -> m ()
+  , narListDir    :: FilePath -> m [FilePath]
+  , narCreateDir  :: FilePath -> m ()
+  , narCreateLink :: FilePath -> FilePath -> m ()
+  , narGetPerms   :: FilePath -> m Directory.Permissions
+  , narSetPerms   :: FilePath -> Directory.Permissions ->  m ()
+  , narIsDir      :: FilePath -> m Bool
+  , narIsSymLink  :: FilePath -> m Bool
+  , narFileSize   :: FilePath -> m Int64
+  , narReadLink   :: FilePath -> m FilePath
+}
+
+
+-- | A particular @NarEffects@ that uses regular POSIX for file manipulation
+--   You would replace this with your own @NarEffects@ if you wanted a
+--   different backend
+narEffectsIO :: IO.MonadIO m => NarEffects m
+narEffectsIO = NarEffects {
+    narReadFile   = IO.liftIO . BSL.readFile
+  , narWriteFile  = \a b -> IO.liftIO $ BSL.writeFile a b
+  , narStreamFile = streamStringOutIO'
+  , narListDir    = IO.liftIO . Directory.listDirectory
+  , narCreateDir  = IO.liftIO . Directory.createDirectory
+  , narCreateLink = \f t -> IO.liftIO $ createSymbolicLink f t
+  , narGetPerms   = IO.liftIO . Directory.getPermissions
+  , narSetPerms   = \f p -> IO.liftIO $ Directory.setPermissions f p
+  , narIsDir      = \d -> fmap isDirectory $ IO.liftIO (getFileStatus d)
+  , narIsSymLink  = IO.liftIO . Directory.pathIsSymbolicLink
+  , narFileSize   = \n -> fmap (fromIntegral . fileSize) $ IO.liftIO (getFileStatus n)
+  , narReadLink   = IO.liftIO . readSymbolicLink
+  }
+
+
+-- | This default implementation for @narStreamFile@ requires @IO.MonadIO@
+streamStringOutIO'
+  :: forall m.IO.MonadIO m
+  => FilePath
+  -> m (Maybe BS.ByteString)
+  -> m ()
+streamStringOutIO' f getChunk = do
+  h <- IO.liftIO $ do
+    IO.openFile f IO.WriteMode
```

Comment:
Bracket?

Reply:
`bracket` from base doesn't like the `MonadIO m =>`, or the recursion. Both issues, could probably be solved (e.g. by changing this function's signature to `FilePath -> IO (Maybe ByteString) -> IO ()`, using `finally`). But what's the goal - making sure we clean up objects that we've created if things go wrong?

Reply:
Yeah, exactly

Reply:
@shlevy Ok, good catch. I wrapped the whole parser/unpacker with a cleanup action here: 0d0c036e9cc170933e490765817d41296d0f8e4c  I tested this manually a couple ways.
 - Modify a NAR file by changing an embedded filepath to one where user doesn't have write permission
 - Modify a NAR file by deleting some of the string padding bytes
I did both of these late in the NAR file, so that unpacking first encountered valid data. In the first case, we get a filesystem exception, in the second we get a `Left` from the parser. In both cases, the cleanup action cleans up as expected.

Reply:
Hmm if this is an exposed API (which it seems like it is as part of `narEffectsIO`), I'd expect it to be async-exception safe.

Reply:
Yeah, that's fair.

Reply:
Addressed now in 98d773696c4f5593b868a55fc29d8c39945cf0b4

Reply:
This doesn't work, you still have an un-bracketed handle before the loop. Any reason not to just do a normal bracket here?

Reply:
Maybe we should do a paired code review?

Reply:
How's 93d2ed5f065c966d02b317d7867da4327f554f76 look, Mr. Shea? I'm open to a paired code review if you want to give me a call.

Reply:
I'd just overlooked the `handle` (only thought about cleaning up the file). We always want to cleanup the handle, but only want to cleanup the file if there's an exception, hence the combination of `bracket` and `catch`. Seems reasonable?

Reply:
Yep!